routing/index_graph.cpp

```diff
@@ -41,62 +39,30 @@ void IndexGraph::Import(vector<Joint> const & joints)
   Build(static_cast<uint32_t>(joints.size()));
 }
 
-Joint::Id IndexGraph::InsertJoint(RoadPoint const & rp)
+double IndexGraph::CalcSegmentWeight(Segment const & segment)
 {
-  Joint::Id const existId = m_roadIndex.GetJointId(rp);
-  if (existId != Joint::kInvalidId)
-    return existId;
-
-  Joint::Id const jointId = m_jointIndex.InsertJoint(rp);
-  m_roadIndex.AddJoint(rp, jointId);
-  return jointId;
+  return m_estimator->CalcSegmentWeight(segment, m_geometry.GetRoad(segment.GetFeatureId()));
 }
 
-bool IndexGraph::JointLiesOnRoad(Joint::Id jointId, uint32_t featureId) const
-{
-  bool result = false;
-  m_jointIndex.ForEachPoint(jointId, [&result, featureId](RoadPoint const & rp) {
-    if (rp.GetFeatureId() == featureId)
-      result = true;
-  });
-
-  return result;
-}
-
-void IndexGraph::GetNeighboringEdges(RoadPoint const & rp, bool isOutgoing,
-                                     vector<JointEdge> & edges)
+void IndexGraph::GetNeighboringEdges(Segment const & from, RoadPoint const & rp, bool isOutgoing,
+                                     vector<SegmentEdge> & edges)
 {
   RoadGeometry const & road = m_geometry.GetRoad(rp.GetFeatureId());
-
   bool const bidirectional = !road.IsOneWay();
-  if (!isOutgoing || bidirectional)
-    GetNeighboringEdge(road, rp, false /* forward */, edges);
 
-  if (isOutgoing || bidirectional)
-    GetNeighboringEdge(road, rp, true /* forward */, edges);
-}
+  if ((isOutgoing || bidirectional) && rp.GetPointId() + 1 < road.GetPointsCount())
+    GetNeighboringEdge(from, Segment(rp.GetFeatureId(), rp.GetPointId(), isOutgoing), isOutgoing,
+                       edges);
 
-void IndexGraph::GetNeighboringEdge(RoadGeometry const & road, RoadPoint const & rp, bool forward,
-                                    vector<JointEdge> & edges) const
-{
-  pair<Joint::Id, uint32_t> const & neighbor = m_roadIndex.FindNeighbor(rp, forward);
-  if (neighbor.first != Joint::kInvalidId)
-  {
-    double const distance =
-        m_estimator->CalcEdgesWeight(rp.GetFeatureId(), road, rp.GetPointId(), neighbor.second);
-    edges.push_back({neighbor.first, distance});
-  }
+  if ((!isOutgoing || bidirectional) && rp.GetPointId() > 0)
+    GetNeighboringEdge(from, Segment(rp.GetFeatureId(), rp.GetPointId() - 1, !isOutgoing),
```

Comment:
Okay, seems that I missed that the calculations of rp depend on isOutgoing. Rechecking now...